src/java/org/apache/cassandra/db/compaction/SharedCompactionObserver.java

```diff
@@ -34,14 +42,28 @@
 /// if assertions are enabled.
 public class SharedCompactionObserver implements CompactionObserver
 {
+    private static final Logger logger = LoggerFactory.getLogger(SharedCompactionObserver.class);
+
     private final AtomicInteger toReportOnComplete = new AtomicInteger(0);
     private final AtomicReference<Throwable> onCompleteException = new AtomicReference(null);
     private final AtomicReference<CompactionProgress> inProgressReported = new AtomicReference<>(null);
-    private final CompactionObserver observer;
+
+    private final List<CompactionObserver> compObservers;
 
     public SharedCompactionObserver(CompactionObserver observer)
     {
-        this.observer = observer;
+        this(observer, null);
+    }
+
+    public SharedCompactionObserver(CompactionObserver primary, @Nullable CompactionObserver secondary)
+    {
+        if (primary == null)
+            throw new IllegalArgumentException("Primary observer cannot be null");
+
+        this.compObservers = new ArrayList<>(2);
+        this.compObservers.add(primary);
+        if (secondary != null)
+            this.compObservers.add(secondary);
```

Comment:
For proper publishing to other threads, prepare the array before setting the final field.

And also make it unmodifiable (e.g. `this.compObservers = secondary != null ? ImmutableList.of(primary, secondary) : ImmutableList.of(primary)`).

Reply:
fixed in https://github.com/datastax/cassandra/pull/1774/commits/6cf01dd6a3d23bb7c5a9ef787b4b6e49f5ed6689